File: src/backend/python/az/api/middleware/error_handler.py
from fastapi import Request
from fastapi.responses import JSONResponse
from fastapi.exceptions import RequestValidationError
from starlette.exceptions import HTTPException as StarletteHTTPException

from api.core.exceptions import AzureAPIException
from api.core.logging import get_logger, correlation_id
from api.models.resource_group import ErrorResponse, ErrorDetail

logger = get_logger(__name__)


async def azure_exception_handler(request: Request, exc: AzureAPIException) -> JSONResponse:
    """Handle Azure API specific exceptions."""
    
    logger.warning(f"Azure API exception: {exc.detail}")
    
    error_response = ErrorResponse(
        error=exc.__class__.__name__,
        message=exc.detail,
        correlation_id=exc.correlation_id or correlation_id.get() or "unknown"
    )
    
    return JSONResponse(
        status_code=exc.status_code,
        content=error_response.model_dump()
    )


async def validation_exception_handler(request: Request, exc: RequestValidationError) -> JSONResponse:
    """Handle request validation errors."""
    
    logger.warning(f"Validation error: {exc.errors()}")
    
    # Convert validation errors to our format
    error_details = [
        ErrorDetail(
            field=".".join(str(loc) for loc in error["loc"]),
            message=error["msg"],
            code=error["type"]
        )
        for error in exc.errors()
    ]
    
    error_response = ErrorResponse(
        error="ValidationError",
        message="Request validation failed",
        details=error_details,
        correlation_id=correlation_id.get() or "unknown"
    )
    
    return JSONResponse(
        status_code=422,
        content=error_response.model_dump()
    )


async def http_exception_handler(request: Request, exc: StarletteHTTPException) -> JSONResponse:
    """Handle general HTTP exceptions."""
    
    logger.warning(f"HTTP exception: {exc.detail}")
    
    error_response = ErrorResponse(
        error="HTTPException",
        message=exc.detail,
        correlation_id=correlation_id.get() or "unknown"
    )
    
    return JSONResponse(
        status_code=exc.status_code,
        content=error_response.model_dump()
    )


async def general_exception_handler(request: Request, exc: Exception) -> JSONResponse:
    """Handle unexpected exceptions."""
    
    logger.error(f"Unexpected error: {str(exc)}", exc_info=True)
    
    error_response = ErrorResponse(
        error="InternalServerError",
        message="An unexpected error occurred",
        correlation_id=correlation_id.get() or "unknown"
    )
    
    return JSONResponse(
        status_code=500,
        content=error_response.model_dump()
    )

File: src/backend/typescript/az/src/middleware/errorHandler.ts
import { Request, Response, NextFunction } from 'express';
import Joi from 'joi';
import { AzureResourceError } from '@/services/azureService';
import { RequestWithCorrelation, ErrorResponse } from '@/types/express';
import { createLogger } from '@/config/logger';

export function errorHandler(
  error: Error,
  req: Request,
  res: Response,
  next: NextFunction
): void {
  const correlationId = (req as RequestWithCorrelation).correlationId || 'unknown';
  const logger = createLogger(correlationId);
  
  let statusCode = 500;
  let code = 'INTERNAL_SERVER_ERROR';
  let message = 'An unexpected error occurred';
  let details: any = undefined;

  // Handle Joi validation errors
  if (Joi.isError(error)) {
    statusCode = 422;
    code = 'VALIDATION_ERROR';
    message = 'Request validation failed';
    details = error.details.map(detail => ({
      field: detail.path.join('.'),
      message: detail.message,
      value: detail.context?.value
    }));
    
    logger.warn('Validation error', { error: error.message, details });
  }
  // Handle Azure-specific errors
  else if (error instanceof AzureResourceError) {
    statusCode = error.statusCode;
    code = error.code;
    message = error.message;
    
    logger.error('Azure resource error', { 
      code: error.code,
      statusCode: error.statusCode,
      message: error.message
    });
  }
  // Handle other known error types
  else if (error.name === 'UnauthorizedError') {
    statusCode = 401;
    code = 'UNAUTHORIZED';
    message = 'Authentication required';
    
    logger.warn('Unauthorized access attempt');
  }
  else if (error.name === 'ForbiddenError') {
    statusCode = 403;
    code = 'FORBIDDEN';
    message = 'Access forbidden';
    
    logger.warn('Forbidden access attempt');
  }
  // Handle unexpected errors
  else {
    logger.error('Unexpected error', { 
      error: error.message,
      stack: error.stack,
      name: error.name
    });
  }

  const errorResponse: ErrorResponse = {
    success: false,
    error: {
      code,
      message,
      details
    },
    correlationId,
    timestamp: new Date().toISOString()
  };

  res.status(statusCode).json(errorResponse);
}

export function notFoundHandler(req: Request, res: Response): void {
  const correlationId = (req as RequestWithCorrelation).correlationId || 'unknown';
  
  const errorResponse: ErrorResponse = {
    success: false,
    error: {
      code: 'NOT_FOUND',
      message: `Route ${req.method} ${req.path} not found`
    },
    correlationId,
    timestamp: new Date().toISOString()
  };

  res.status(404).json(errorResponse);
}